### ISSUE TITLE
ES 2.2.0 wrongfully marks shard as unassigned / corrupted

### ISSUE THREAD

**Issue by jrots** [elastic/elasticsearch#16793]
Hi I had a red state on one of the shards in my index : 

I tried a force relocation to reassign it, this was the output:
https://gist.github.com/jrots/da71bc8b0ad86a27636a
force allocation from previous reason ALLOCATION_FAILED, failed recovery,
 failure IndexShardRecoveryException[failed to fetch index version after copying it over]; nested: 
 CorruptIndexException[failed engine (reason: [search execution corruption failure]) (resource=preexisting_corruption)]; nested: NotSerializableExceptionWrapper[failed engine (reason: [search execution corruption failure])]; 
 nested: CorruptIndexException[Corrupted: docID=3412736, docBase=53, chunkDocs=50, numDocs=4460365
There was indeed a "corrupted_.." file present, 

I did a check with lucene CheckIndex, this was the output : 
https://gist.github.com/jrots/e816c24b17ef4ae6fe29

all segments showed up as valid...  

I manually removed the corrupt_ file and did a restart of that node & all was assigned again and ES had a green state again.. but there might be a bug in ES that generates a "corrupted_" file too soon because there weren't any issues when doing a check afterwards..

**Comment by jrots** [comment]
Ok seems like #16789 is related and I'm no the only one experiencing this

**Comment by s1monw** [comment]
Hey @jrots thanks for reporting this, lemme ask some questions:
 *  do you still have that "corrupt_" file by any chance? 
 * Can you also check you logs to tell what lead to this corruption in the first place? 
 * Did you have replicas before and one of them got corrupted? 
 * Do you know when that corruption file was written vs. when the index files were written?
 * have you done a restore by any chance?

thanks!

**Comment by jrots** [comment]
Hi @s1monw 
- afraid I have deleted the corrupt file :/ If it would reoccur I'll make sure I have a copy of it. 
- log file of when it occurred : https://gist.github.com/MassiveMedia/18014fc75c7fbd6cf0eb
- I didn't have any replicas enabled on that index.. so can't answer on that one 
- no afraid I didn't do an `ls -lsha` at that time, just a simple `ls` .. :/ so can't provide you that info.. will keep this in mind when it would happen again. 
- no I didn't do a restore

**Comment by s1monw** [comment]
@jrots I have the suspicion that this is a compression bug in lucene. Do you see any chance to upload this shards index somewhere so we can investigate.

**Comment by s1monw** [comment]
@jpountz @rmuir @mikemccand just pinging you here to get your attention as well

**Comment by jpountz** [comment]
Hmm that looks like a bug in the index on top of stored fields. Maybe an overflow?

**Comment by jrots** [comment]
Ok I've uploaded the index, but I rather not post the link here as it contains some sensitive data. Can I email the link to simonw at apache dot org?

**Comment by s1monw** [comment]
you can send it to simon at elastic dot co

**Comment by mikemccand** [comment]
It does sounds like maybe a Lucene bug, especially if `CheckIndex` can't find anything wrong with the segment, but a search thread did.  The JVM is modern (1.8.0_66).

**Comment by mikemccand** [comment]
Hi @jrots, I ran `CheckIndex` from both 5.4.x and 5.5.x Lucene on your index, but it did in fact detect a checksum mismatch on that one segment (`_69l`): https://gist.github.com/mikemccand/1be773920096ede16481

Also, this index has 33 segments but your `CheckIndex` output shows 19.

Are you sure you copied up exactly the same index you had successfully run `CheckIndex` on?  Otherwise I can't explain why your `CheckIndex` runs fine but when I run it, it detects corruption.

**Comment by jrots** [comment]
Hi, 
Well the index data was from almost 10 hours after I removed the _corrupted file.. and there have been updates to the data in that index in the meanwhile. So could be that the segment was merged internally already. I also copied the data with a simple `cp` command while elastic was still running, which probably was not the best idea (need to shutdown the node first). 
If I run `CheckIndex` on that copied data I also get the same warning, https://gist.github.com/jrots/c4087c8fab52a81d346d
I made a new backup now of this shard, while ES was shutdown and `CheckIndex` also ran fine on it.. 
but it probably doesn't make sense to send you that data over? 
If I ever hit this issue again, I'll make sure I make a backup of the data at that point. Because atm it's probably hard to investigate what exactly was corrupt.
